File: core/player_selection.py
"""Player selection UI for Pyglet games."""
import time
from typing import List, Dict, Tuple
from config import PLAYER_COLORS, HOVER_TIME_THRESHOLD, Colors


class PlayerSelectionUI:
    """Player selection screen with 8 player slots."""
    
    def __init__(self, screen_width: int, screen_height: int):
        self.screen_width = screen_width
        self.screen_height = screen_height
        self.selected = [False] * 8
        self.hover_states = {}
        self.start_ready = False  # Track if start button was clicked
        
        # Calculate slot positions
        self.slots = self._calculate_slots()
        
        # Start button in center
        self.start_button_pos = (screen_width // 2, screen_height // 2)
        self.start_button_radius = 80
    
    def _calculate_slots(self) -> List[Tuple[int, int, int, int]]:
        """Calculate circular position for each player slot (x, y, radius, radius)."""
        slots = []
        
        # Circular layout matching Pygame
        spacing_x = self.screen_width // 4
        spacing_y = self.screen_height // 3
        slot_radius = 55
        
        # Positions: 3 bottom, 3 top, 1 left, 1 right
        positions = [
            (spacing_x * 1, self.screen_height - spacing_y // 2),  # Bottom left
            (spacing_x * 2, self.screen_height - spacing_y // 2),  # Bottom center
            (spacing_x * 3, self.screen_height - spacing_y // 2),  # Bottom right
            (spacing_x * 1, spacing_y // 2),  # Top left
            (spacing_x * 2, spacing_y // 2),  # Top center
            (spacing_x * 3, spacing_y // 2),  # Top right
            (spacing_x // 2, self.screen_height // 2),  # Left
            (self.screen_width - spacing_x // 2, self.screen_height // 2)  # Right
        ]
        
        for x, y in positions:
            # Return as (x, y, radius, radius) for compatibility
            slots.append((x - slot_radius, y - slot_radius, slot_radius * 2, slot_radius * 2))
        
        return slots
    
    def update_with_fingertips(self, fingertip_meta: List[Dict], min_players: int = 2):
        """Update selection based on fingertip positions."""
        current_time = time.time()
        active_hovers = set()
        
        for meta in fingertip_meta:
            pos = meta["pos"]
            is_click = bool(meta.get("click"))
            
            # Check start button first (if enough players selected)
            if self.selected_count() >= min_players:
                cx, cy = self.start_button_pos
                dist = ((pos[0] - cx) ** 2 + (pos[1] - cy) ** 2) ** 0.5
                
                if dist <= self.start_button_radius:
                    if is_click:
                        self.start_ready = True
                        continue
                    key = "start_button"
                    active_hovers.add(key)
                    
                    if key not in self.hover_states:
                        self.hover_states[key] = {"start_time": current_time, "pos": pos}
                    
                    hover_duration = current_time - self.hover_states[key]["start_time"]
                    if hover_duration >= HOVER_TIME_THRESHOLD:
                        self.start_ready = True
                        del self.hover_states[key]
                    continue
            
            # Check player slots
            for i, (x, y, w, h) in enumerate(self.slots):
                # Calculate center and check circular distance
                cx = x + w // 2
                cy = y + h // 2
                radius = w // 2
                dist = ((pos[0] - cx) ** 2 + (pos[1] - cy) ** 2) ** 0.5
                
                if dist <= radius:
                    if is_click:
                        if i != 7:
                            self.selected[i] = not self.selected[i]
                        break
                    key = f"slot_{i}"
                    active_hovers.add(key)
                    
                    if key not in self.hover_states:
                        self.hover_states[key] = {"start_time": current_time, "pos": pos}
                    
                    hover_duration = current_time - self.hover_states[key]["start_time"]
                    if hover_duration >= HOVER_TIME_THRESHOLD:
                        # Don't allow deselecting Player 8 (DM)
                        if i != 7:
                            self.selected[i] = not self.selected[i]
                        del self.hover_states[key]
                    break
        
        # Remove stale hover states
        for key in list(self.hover_states.keys()):
            if key not in active_hovers:
                del self.hover_states[key]
    
    def get_hover_progress(self) -> List[Dict]:
        """Get hover progress for each active slot."""
        current_time = time.time()
        progress_list = []
        
        for key, state in self.hover_states.items():
            slot_idx = int(key.split("_")[1])
            hover_duration = current_time - state["start_time"]
            progress = min(1.0, hover_duration / HOVER_TIME_THRESHOLD)
            x, y, w, h = self.slots[slot_idx]
            progress_list.append({
                "pos": (x, y),
                "progress": progress,
                "slot": slot_idx
            })
        
        return progress_list
    
    def selected_count(self) -> int:
        """Get number of selected players."""
        return sum(self.selected)
    
    def get_selected_indices(self) -> List[int]:
        """Get list of selected player indices."""
        return [i for i, selected in enumerate(self.selected) if selected]
    
    def reset(self):
        """Reset all selections."""
        self.selected = [False] * 8
        self.hover_states = {}
        self.start_ready = False
    
    def closest_player_color(self, pos: Tuple[int, int]) -> Tuple[int, int, int]:
        """Get the closest player color for a cursor position."""
        min_dist = float('inf')
        closest_idx = 0
        
        for i, (x, y, w, h) in enumerate(self.slots):
            cx = x + w // 2
            cy = y + h // 2
            dist = ((pos[0] - cx) ** 2 + (pos[1] - cy) ** 2) ** 0.5
            if dist < min_dist:
                min_dist = dist
                closest_idx = i
        
        return PLAYER_COLORS[closest_idx]

File: ARPi2.Sharp/ARPi2.Sharp/Core/PlayerSelection.cs
using System;
using System.Collections.Generic;

namespace ARPi2.Sharp.Core;

/// <summary>
/// Player selection screen with 8 player slots in circular layout.
/// Port of Python core/player_selection.py.
/// </summary>
public class PlayerSelectionUI
{
    public int ScreenWidth, ScreenHeight;
    public bool[] Selected = new bool[8];
    public bool StartReady;

    private readonly (int X, int Y, int R)[] _slots; // centre + radius
    private readonly Dictionary<string, float> _hoverStarts = new();
    public (int X, int Y) StartButtonPos;
    public int StartButtonRadius = 80;

    public PlayerSelectionUI(int screenW, int screenH)
    {
        ScreenWidth = screenW; ScreenHeight = screenH;
        int sx = screenW / 4, sy = screenH / 3;
        int sr = 55;
        StartButtonPos = (screenW / 2, screenH / 2);

        _slots = new (int, int, int)[]
        {
            (sx * 1, screenH - sy / 2, sr),
            (sx * 2, screenH - sy / 2, sr),
            (sx * 3, screenH - sy / 2, sr),
            (sx * 1, sy / 2, sr),
            (sx * 2, sy / 2, sr),
            (sx * 3, sy / 2, sr),
            (sx / 2, screenH / 2, sr),
            (screenW - sx / 2, screenH / 2, sr),
        };
    }

    public int SelectedCount() { int n = 0; foreach (bool s in Selected) if (s) n++; return n; }
    public List<int> GetSelectedIndices()
    {
        var list = new List<int>();
        for (int i = 0; i < 8; i++) if (Selected[i]) list.Add(i);
        return list;
    }

    public void Reset() { Array.Fill(Selected, false); _hoverStarts.Clear(); StartReady = false; }

    public void UpdateWithInputs(List<(int X, int Y, bool Click)> inputs, float currentTime, int minPlayers = 2)
    {
        var active = new HashSet<string>();

        foreach (var (ix, iy, click) in inputs)
        {
            // Start button
            if (SelectedCount() >= minPlayers)
            {
                var (scx, scy) = StartButtonPos;
                float dist = MathF.Sqrt((ix - scx) * (ix - scx) + (iy - scy) * (iy - scy));
                if (dist <= StartButtonRadius)
                {
                    if (click) { StartReady = true; continue; }
                    string key = "start";
                    active.Add(key);
                    if (!_hoverStarts.ContainsKey(key)) _hoverStarts[key] = currentTime;
                    if (currentTime - _hoverStarts[key] >= GameConfig.HoverTimeThreshold)
                    {
                        StartReady = true;
                        _hoverStarts.Remove(key);
                    }
                    continue;
                }
            }

            // Player slots
            for (int i = 0; i < 8; i++)
            {
                var (scx, scy, sr) = _slots[i];
                float dist = MathF.Sqrt((ix - scx) * (ix - scx) + (iy - scy) * (iy - scy));
                if (dist > sr) continue;

                if (click) { if (i != 7) Selected[i] = !Selected[i]; break; }
                string key = $"slot_{i}";
                active.Add(key);
                if (!_hoverStarts.ContainsKey(key)) _hoverStarts[key] = currentTime;
                if (currentTime - _hoverStarts[key] >= GameConfig.HoverTimeThreshold)
                {
                    if (i != 7) Selected[i] = !Selected[i];
                    _hoverStarts.Remove(key);
                }
                break;
            }
        }

        // Remove stale hovers
        var stale = new List<string>();
        foreach (var k in _hoverStarts.Keys) if (!active.Contains(k)) stale.Add(k);
        foreach (var k in stale) _hoverStarts.Remove(k);
    }

    public void Draw(Renderer r)
    {
        // Draw slots
        for (int i = 0; i < 8; i++)
        {
            var (cx, cy, sr) = _slots[i];
            var col = GameConfig.PlayerColors[i];
            if (Selected[i])
            {
                r.DrawCircle(col, (cx, cy), sr, alpha: 200);
                r.DrawCircle((255, 255, 255), (cx, cy), sr, width: 3, alpha: 255);
            }
            else
            {
                r.DrawCircle((60, 60, 70), (cx, cy), sr, alpha: 120);
                r.DrawCircle(col, (cx, cy), sr, width: 2, alpha: 180);
            }
            r.DrawText($"P{i + 1}", cx, cy, fontSize: 22, color: (255, 255, 255),
                       anchorX: "center", anchorY: "center", bold: true);
        }

        // Start button
        if (SelectedCount() >= GameConfig.MinPlayers)
        {
            var (sx, sy) = StartButtonPos;
            r.DrawCircle((40, 180, 60), (sx, sy), StartButtonRadius, alpha: 220);
            r.DrawCircle((255, 215, 0), (sx, sy), StartButtonRadius, width: 3);
            r.DrawText("START", sx, sy, fontSize: 28, color: (255, 255, 255),
                       anchorX: "center", anchorY: "center", bold: true);
        }
    }
}
